### PR TITLE
Initial implementation of Rules Engine

### PR DESCRIPTION
**Pull request by makyo**
This PR is for an initial implementation of the Replay Rules Engine, intended to replace the Exclusion oracle.

It's a 'fake' PR, in that it's a merge of master into a throw-away branch, but this will allow folks to comment on code easily.

### CODE REVIEW

File: rulesengine/rules/tests/test_models.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+                'collection': 'Planets',
+                'partner': 'Holst',
+                'warc_match': 'jupiter',
+                'rewrite_from': 'zeus',
+                'rewrite_to': 'jupiter',
+                'public_comment': 'initial creation',
+                'private_comment': 'going roman',
+                'enabled': True,
+            },
+        })
```
Comment (jkafader) [inline rulesengine/rules/tests/test_models.py]: 👍

File: rulesengine/rules/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            {
+                'surt': 'https://(org,archive,',
+                'capture-date': 'bad-wolf',
+            })
+        self.assertEqual(response.status_code, 400)
+        parsed = json.loads(response.content.decode('utf-8'))
+        self.assertEqual(parsed['status'], 'error')
+        self.assertEqual(
+            parsed['message'], 'capture-date query string param must be '
+            'ISO 8601 format')
```
Comment (jkafader) [inline rulesengine/rules/tests/test_views.py]: 👍

File: rulesengine/rules/utils/json.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    result = {
+        'status': 'error',
+        'message': message,
+    }
+    if obj is not None:
+        result['result'] = obj
+    return HttpResponse(ujson.dumps(result, pre_encode_hook=date_renderer),
+                        content_type='application/json',
+                        status=400)
```
Comment (jkafader) [inline rulesengine/rules/utils/json.py]: these are interesting helpers -- I was wondering if 400 is always appropriate. I guess it is at the moment.

File: rulesengine/rules/utils/surt.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        Arguments:
+        url -- The URL to SURTify.
+
+        Returns:
+        A SURT broken down into its parts.
+        """
+        return cls(parse_url(url).surt().decode('utf-8'))
+
+    def __str__(self):
+        return self.surt
```
Comment (jkafader) [inline rulesengine/rules/utils/surt.py]: I don't remember if we decided to use an external `surt` library or not. Let's talk about this.

File: rulesengine/rules/utils/tests/test_json.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    def test_error_with_obj(self):
+        response = json.error("dalek", {'exterminate': True})
+        expected = {
+            'status': 'error',
+            'message': 'dalek',
+            'result': {'exterminate': True},
+        }
+        self.assertEqual(
+            ujson.loads(response.content.decode()),
+            expected)
```
Comment (jkafader) [inline rulesengine/rules/utils/tests/test_json.py]: 👍

File: rulesengine/rules/utils/tests/test_surt.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        surt = Surt.from_url(
+            'https://example.com/doctor/companion?name=rose#actress')
+        self.assertEqual(surt.parts, [
+            'https://(', 'com,', 'example,', ')', '/doctor', '/companion',
+            '?name=rose', '#actress'])
+
+    def test_str(self):
+        given = 'http://(org,archive,)'
+        surt = Surt(given)
+        self.assertEqual(str(surt), given)
```
Comment (jkafader) [inline rulesengine/rules/utils/tests/test_surt.py]: 👍 (depends on comments above)

File: rulesengine/rules/utils/tests/test_tree.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            results, ['BLOCK PLAYBACK (https://(com,example,)/partner)'])
+
+    def test_capture_date(self):
+        # The following rule blocks content on April Fool's Day
+        date = datetime(2018, 4, 1, 12, 0, 0, 0, timezone.utc)
+        results = [str(rule) for rule in
+                   tree(
+                       Surt('https://(com,example,)/blocked'),
+                       capture_date=date)]
+        self.assertEqual(results, ['BLOCK PLAYBACK (https://(com,example,))'])
```
Comment (jkafader) [inline rulesengine/rules/utils/tests/test_tree.py]: 👍

File: rulesengine/rules/utils/tests/test_validators.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                'retrieve_date': {
+                    'start': self.now.isoformat(),
+                    'end': 'bad-wolf 4',
+                },
+            })
+        self.assertEqual(
+            str(context.exception),
+            str(ValueError((
+                'retrieve end date',
+                ValueError('Unknown string format:', 'bad-wolf 4')))))
```
Comment (jkafader) [inline rulesengine/rules/utils/tests/test_validators.py]: 👍

File: rulesengine/rules/utils/tree.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        filters = filters & Q(collection=collection)
+    if partner is not None:
+        filters = filters & Q(partner=partner)
+    if capture_date is not None:
+        filters = filters & (
+            Q(capture_date_start__isnull=False) &
+            Q(capture_date_end__isnull=False) &
+            Q(capture_date_start__lt=capture_date) &
+            Q(capture_date_end__gt=capture_date))
+    return Rule.objects.filter(filters)
```
Comment (jkafader) [inline rulesengine/rules/utils/tree.py]: complicated -- let's talk about this in more detail

File: rulesengine/rules/utils/validators.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            raise ValueError(('capture end date', e))
+    if 'retrieve_date' in input:
+        try:
+            parse_date(input['retrieve_date']['start'])
+        except ValueError as e:
+            raise ValueError(('retrieve start date', e))
+        try:
+            parse_date(input['retrieve_date']['end'])
+        except ValueError as e:
+            raise ValueError(('retrieve end date', e))
```
Comment (jkafader) [inline rulesengine/rules/utils/validators.py]: 👍

File: rulesengine/rules/views.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            'capture-date query string param must be '
+            'ISO 8601 format', str(e))
+    surt = Surt(surt_qs)
+    tree_result = tree(
+        surt,
+        neg_surt=request.GET.get('neg-surt'),
+        collection=request.GET.get('collection'),
+        partner=request.GET.get('partner'),
+        capture_date=capture_date)
+    return success([rule.summary() for rule in tree_result])
```
Comment (jkafader) [inline rulesengine/rules/views.py]: 👍 The logic for PUT/POST appears correct now, looks good overall

File: rulesengine/rulesengine/settings.py [hunk 12]
```diff
@@ -37,6 +37,8 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'django_extensions',
+    'rules',
 ]
 
 MIDDLEWARE = [
```
Comment (jkafader) [inline rulesengine/rulesengine/settings.py]: Is all of this middleware necessary? might speed requests slightly if we take out whatever we don't need (e.g. csrf, message, xframeoptions, maybe common?

File: rulesengine/rulesengine/urls.py [hunk 13]
```diff
@@ -16,6 +16,13 @@
...
+from rules import views
+
+
 urlpatterns = [
     path('admin/', admin.site.urls),
+    path('rules', views.RulesView.as_view()),
+    path('rules/tree/<path:surt_string>', views.tree_for_surt),
+    path('rules/for-request', views.rules_for_request),
+    path('rule/<int:pk>', views.RuleView.as_view()),
 ]
```
Comment (jkafader) [inline rulesengine/rulesengine/urls.py]: nicely done urls, appreciated the type hints; good names.

**Comment by makyo** [comment]
Agreed this is a good step over voice.

**Pull request closed without merging**